### PR TITLE
🐛 Fix AbstractNode#getRole

### PR DESCRIPTION
**Pull request by bitPogo**
see #1066

### CODE REVIEW

**Comment by bitPogo** [comment]
Test are missing...I am not sure where to put them (under types). Also I made a output from get role and currently it returns a object, if I use a plain blocknode.

**Comment by bitPogo** [comment]
Odd..I cannot reproduce the error (of the test) locally, but I get another one: 

```bash
 1) Node.js
       Node.js
         Loading
           should get the line of number of a block when sourcemap is enabled:
     AssertionError: expected undefined to equal 3
      at Context.<anonymous> (spec/share/asciidoctor-spec.js:360:48)
      at processImmediate (internal/timers.js:461:21)
```

**Comment by Mogztter** [comment]
> Odd..I cannot reproduce the error (of the test) locally, but I get another one:

This error was fixed on master (https://github.com/asciidoctor/asciidoctor.js/commit/0d00235b6e473f51cf835f8cd627c4ef9675f67b) maybe your branch is a bit out of sync?

Anyway, you should test locally against a release because we are currently merging a few changes on the master branch of Asciidoctor (Ruby) that might impact Asciidoctor.js.

You can run (at the root of this project):

```
$ npm run clean --prefix packages/core
$ ASCIIDOCTOR_CORE_VERSION=v2.0.10 npm run build --prefix packages/core
```

**Comment by bitPogo** [comment]
That does the trick - probably the some tingeling with a cache.

File: packages/core/src/asciidoctor-core-api.js [hunk 1]
```diff
@@ -1035,7 +1035,8 @@ AbstractNode.prototype.isRole = function (expectedValue) {
  * @memberof AbstractNode
  */
 AbstractNode.prototype.getRole = function () {
-  return this.$role()
+  const role = this.$role()
+  return Opal.nil ? undefined : role
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js:1039]: I think it should be:

```suggestion
  return role === Opal.nil ? undefined : role
```

a test will confirm that :wink:

**Comment by Mogztter** [comment]
> That does the trick - probably the some tingeling with a cache.

We are using an aggressive cache because transpiling Asciidoctor Ruby to JavaScript (using Opal) is somewhat "slow".

File: packages/core/src/asciidoctor-core-api.js (hunk 1 above)
Comment (bitPogo) [reply packages/core/src/asciidoctor-core-api.js:1039]: Have you any idea where to put a test for it...I mean to test in test for types seems a bit weird...but would be do for now...maybe to restructure/split in general test.ts would be a good idea for more structure. So what are your thoughts?

Comment (Mogztter) [reply packages/core/src/asciidoctor-core-api.js]: > Have you any idea where to put a test for it...I mean to test in test for types seems a bit weird...but would be do for now...maybe to restructure/split in general test.ts would be a good idea for more structure. So what are your thoughts?

Yes, at some point, we should `tests.ts`, the file is starting to get quite long but we should do that in a dedicated pull request.

The idea is to add a test in `tests.ts` to make sure that the type definition is correct.

You can add a test in `packages/core/spec/share/asciidoctor-spec.js`.
The idea is to load a document and call `getRole` on a block that does not have a role:

_untested-code.js_
```js
const doc = asciidoctor.load(`= Title

== Section without role
`)
const sectionWithoutRole = doc.getBlocks()[0]
expect(sectionWithoutRole.getRole()).to.be.undefined()
```

All tests in "share/asciidoctor-spec.js" will be executed against all supported runtime, currently:
- Node
- Browser
- GraalVM

**Comment by Mogztter** [comment]
@bitPogo Let me know if it does make sense, if not I will try my best to clarify :wink: 
The code looks good but we need to add a few tests to make sure that everything works as intended and that we won't break the code in the future.

**Comment by bitPogo** [comment]
I already tinkered with some test - however I do not get the expected return value (`string`). I actually get a `object` which looks like:

```javascript
constructor {
        '$$id': 4,
        apply: [Function (anonymous)],
        call: [Function (anonymous)]
}
```

**Comment by Mogztter** [comment]
This object is `Opal.nil`, did you run clean and build?

**Comment by Mogztter** [comment]
@bitPogo I hope you don't mind, I've added tests and I will merge this pull request as soon as the build is green :tada:

**Comment by Mogztter** [comment]
Actually I will let this pull request open a bit longer so you get a chance to review my changes :wink:

**Review by bitPogo — COMMENTED** [review]
This looks pretty good, however, I currently have 2 project, where `getRole` returns something like I mentions in the discussion. After the release I will take a 2nd look in the projects and try to pin point the problem and report back.

**Comment by Mogztter** [comment]
And merged, thanks @bitPogo :+1:

**Pull request merged**